From 74718f6b381932474a6b9df5cbfa6c9e04d599d2 Mon Sep 17 00:00:00 2001
From: "carlosgc@webkit.org"
 <carlosgc@webkit.org@268f45cc-cd09-0410-ab3c-d52691b4dbfc>
Date: Mon, 6 Feb 2017 18:00:52 +0000
Subject: [PATCH] [Soup] Deadlock in NetworkProcess
 https://bugs.webkit.org/show_bug.cgi?id=167876

Reviewed by Michael Catanzaro.

WebKitSoupRequestInputStream uses a read lock. What is happening is that webkitSoupRequestInputStreamAddData
takes the lock, and it calls webkitSoupRequestInputStreamPendingReadAsyncComplete with the lock help. That
causes webkitSoupRequestInputStreamReadAsync to be called again to read the next chunk, but in the same run loop
operation. We don't really need the read lock because both webkitSoupRequestInputStreamAddData and
webkitSoupRequestInputStreamReadAsync shoudl always be called from the main thread.

* WebProcess/soup/WebKitSoupRequestInputStream.cpp:
(webkitSoupRequestInputStreamReadAsync): Remove the read lock and assert if called from a secondary thread.
(webkitSoupRequestInputStreamAddData): Ditto.

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@211734 268f45cc-cd09-0410-ab3c-d52691b4dbfc
---
 .../WebProcess/soup/WebKitSoupRequestInputStream.cpp    | 11 ++++-------
 2 files changed, 21 insertions(+), 7 deletions(-)

diff --git a/Source/WebKit2/WebProcess/soup/WebKitSoupRequestInputStream.cpp b/Source/WebKit2/WebProcess/soup/WebKitSoupRequestInputStream.cpp
index a3cdc8d..78d9761 100644
--- a/Source/WebKit2/WebProcess/soup/WebKitSoupRequestInputStream.cpp
+++ b/Source/WebKit2/WebProcess/soup/WebKitSoupRequestInputStream.cpp
@@ -20,8 +20,7 @@
 #include "config.h"
 #include "WebKitSoupRequestInputStream.h"
 
-#include <wtf/Lock.h>
-#include <wtf/Threading.h>
+#include <wtf/MainThread.h>
 #include <wtf/glib/GRefPtr.h>
 #include <wtf/glib/GUniquePtr.h>
 
@@ -45,7 +44,6 @@ struct _WebKitSoupRequestInputStreamPrivate {
 
     GUniquePtr<GError> error;
 
-    Lock readLock;
     std::unique_ptr<AsyncReadData> pendingAsyncRead;
 };
 
@@ -85,11 +83,10 @@ static bool webkitSoupRequestInputStreamIsWaitingForData(WebKitSoupRequestInputS
 
 static void webkitSoupRequestInputStreamReadAsync(GInputStream* inputStream, void* buffer, gsize count, int /*priority*/, GCancellable* cancellable, GAsyncReadyCallback callback, gpointer userData)
 {
+    ASSERT(isMainThread());
     WebKitSoupRequestInputStream* stream = WEBKIT_SOUP_REQUEST_INPUT_STREAM(inputStream);
     GRefPtr<GTask> task = adoptGRef(g_task_new(stream, cancellable, callback, userData));
 
-    LockHolder locker(stream->priv->readLock);
-
     if (!webkitSoupRequestInputStreamHasDataToRead(stream) && !webkitSoupRequestInputStreamIsWaitingForData(stream)) {
         g_task_return_int(task.get(), 0);
         return;
@@ -149,11 +146,11 @@ GInputStream* webkitSoupRequestInputStreamNew(uint64_t contentLength)
 
 void webkitSoupRequestInputStreamAddData(WebKitSoupRequestInputStream* stream, const void* data, size_t dataLength)
 {
+    ASSERT(isMainThread());
+
     if (webkitSoupRequestInputStreamFinished(stream))
         return;
 
-    LockHolder locker(stream->priv->readLock);
-
     if (dataLength) {
         // Truncate the dataLength to the contentLength if it's known.
         if (stream->priv->contentLength && stream->priv->bytesReceived + dataLength > stream->priv->contentLength)
-- 
2.9.3

